src/spec-lang/tc/typecheck.ts

```diff
@@ -1962,97 +1962,124 @@ export function tcForAll(expr: SForAll, ctx: STypingCtx, typeEnv: TypeEnv): Type
     const callee = expr.callee;
 
-    if (callee instanceof SId && callee.name === ScribbleBuiltinFunctions.unchecked_sum) {
-        callee.defSite = "builtin_fun";
+    if (callee instanceof SId) {
+        if (callee.name === ScribbleBuiltinFunctions.unchecked_sum) {
+            callee.defSite = "builtin_fun";
 
-        if (expr.args.length !== 1) {
-            throw new SExprCountMismatch(
-                `Calls to sum expect a single argument, not ${expr.args.length} in ${expr.pp()}`,
-                expr
-            );
-        }
+            if (expr.args.length !== 1) {
+                throw new SExprCountMismatch(
+                    `Calls to sum expect a single argument, not ${
+                        expr.args.length
+                    } in ${expr.pp()}`,
+                    expr
+                );
+            }
+
+            const argT = tc(expr.args[0], ctx, typeEnv);
 
-        const argT = tc(expr.args[0], ctx, typeEnv);
+            if (!(argT instanceof PointerType)) {
+                throw new SWrongType(
+                    `sum expects a numeric array or map to numbers, not ${argT.pp()} in ${expr.pp()}`,
+                    expr,
+                    argT
+                );
+            }
+
+            if (argT.to instanceof MappingType && argT.to.valueType instanceof IntType) {
+                return new IntType(256, argT.to.valueType.signed);
+            }
+
+            if (argT.to instanceof ArrayType && argT.to.elementT instanceof IntType) {
+                return new IntType(256, argT.to.elementT.signed);
+            }
 
-        if (!(argT instanceof PointerType)) {
             throw new SWrongType(
                 `sum expects a numeric array or map to numbers, not ${argT.pp()} in ${expr.pp()}`,
                 expr,
                 argT
             );
         }
 
-        if (argT.to instanceof MappingType && argT.to.valueType instanceof IntType) {
-            return new IntType(256, argT.to.valueType.signed);
-        }
+        if (callee.name === ScribbleBuiltinFunctions.eq_encoded) {
+            callee.defSite = "builtin_fun";
 
-        if (argT.to instanceof ArrayType && argT.to.elementT instanceof IntType) {
-            return new IntType(256, argT.to.elementT.signed);
+            if (expr.args.length !== 2) {
+                throw new SExprCountMismatch(
+                    `Calls to ${callee.name} expect a 2 arguments, not ${
+                        expr.args.length
+                    } in ${expr.pp()}`,
+                    expr
+                );
+            }
+
+            /**
+             * @todo Consider checking that type is encodable.
+             * Here is how compiler does it:
+             * https://github.com/ethereum/solidity/blob/0a0c389541c0f247691951edd64451be7145436c/libsolidity/ast/Types.cpp#L314-L339
+             */
+            tc(expr.args[0], ctx, typeEnv);
```

Comment:
Here we need to check that both argument types are 'abi-encodable'. Maybe we can add a new function `isTypeABIEncodable(t: TypeNode): bool` that follows the logic of `toABIEncdodedType`

Reply:
See https://github.com/ConsenSys/solc-typed-ast/issues/206